tests/test_otel.py

```diff
@@ -0,0 +1,524 @@
+            assert scheme == "https", "Only https:// URLs are supported."
+            s.quic_cache_layer.add_domain("127.0.0.1", port)
+        resp = s.get(
+            f"{scheme}://127.0.0.1:{port}{path}", headers=headers, verify=False
+        )
+    collect_headers(resp.headers, http_ver)
+    return resp.text
+
+
+@pytest.mark.usefixtures("trace_service", "_otelcol", "nginx")
+@pytest.mark.parametrize(
+    ("nginx_config", "http_ver", "scheme"),
+    [
+        ({"name": "test_http0", "mode": ""}, 0, "http"),
+        ({"name": "test_http1", "mode": ""}, 1, "http"),
+        ({"name": "test_http2", "mode": "ssl http2"}, 2, "https"),
+        ({"name": "test_http3", "mode": "quic"}, 3, "https"),
+    ],
+    indirect=["nginx_config"],
+    ids=["http 0.9", "http 1.1", "http 2.0 ssl", "http 3.0 quic"],
+    scope="module",
+)
+class TestOTelGenerateSpans:
+    @classmethod
+    def teardown_class(cls):
+        time.sleep(3)  # wait for sending the last batch to collector
+
+    @pytest.mark.parametrize(
+        "headers", [None, context], ids=["no context", "context"]
+    )
+    @pytest.mark.parametrize(
+        ("path", "value"),
+        [
+            ("/", "TRACE-ON"),
+            ("/context-ignore", ""),
+            ("/context-extract", ""),
+            ("/context-inject", ""),
+            ("/context-propagate", ""),
+        ],
+        ids=[
+            "trace on",
+            "context ignore",
+            "context extract",
+            "context inject",
+            "context propagate",
+        ],
+    )
+    def test_make_batch0(
+        self, logger, response, http_ver, scheme, path, headers, value
+    ):
+        assert response == value
+
+    @pytest.mark.parametrize(
+        ("path", "value", "headers"),
+        [("/", "TRACE-ON", None)] * 10,
+        ids=["trace on"] * 10,
+    )
+    def test_make_batch1(
+        self, logger, response, http_ver, scheme, path, headers, value
+    ):
+        assert response == value
+
+    @pytest.mark.parametrize(
+        ("path", "value", "headers"),
+        [("/", "TRACE-ON", None)] * 10,
+        ids=["trace on"] * 10,
+    )
+    def test_make_batch2(
+        self, logger, response, http_ver, scheme, path, headers, value
+    ):
+        assert response == value
+
+    @pytest.mark.parametrize(
+        "headers", [None, context], ids=["no context", "context"]
+    )
+    @pytest.mark.parametrize(
+        ("path", "value"),
+        [("/204", "")],
+        ids=["trace off"],
+    )
+    def test_do_request(
+        self, logger, response, http_ver, scheme, path, headers, value
+    ):
+        assert response == value
+
+
+@pytest.mark.parametrize(
+    "http_ver",
+    [0, 1, 2, 3],
+    ids=["http 0.9", "http 1.1", "http 2.0 ssl", "http 3.0 quic"],
+)
+class TestOTelSpans:
+    @pytest.mark.parametrize(
+        ("idx", "value"), enumerate([10] * 3), ids=["batch"] * 3
+    )
+    def test_batch_size(self, http_ver, batches, idx, value):
```

Comment:
As far as batches are concerned, we need only several tests:
- send 1 full batch of size greater then 1 (say 3)
- send maximum amount of full batches (say 3 by 3)
- send more then maximum amount of batches: some spans might be dropped here, so we can assert that at least maximum amount of spans went through.

All the remaining tests should mostly operate with single request.

Reply:
Fixed.